### PR TITLE
Feature/application app roles

### LINKED ISSUE #75: azuread_application: Add support for custom app roles
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

This is slightly different than the existing requests (that I can see).  This request is for the ability to add custom app roles to an Azure AD App Registration.  This can be done programmatically via PowerShell currently, it would be great to have Terraform support.

### New or Affected Resource(s)

* azuread_application
* azuread_application_group_role
* azuread_application_user_role

### Potential Terraform Configuration

```hcl
data "azuread_application_group_role" "myCustomGroupRole" {
    name = "myCustomRole"
    azureADGroupName = "MyAzureADGroupName"
    }

data "azuread_application_user_role "myCustomUserRole" {
    name = "myCustomUserRole"
    azureADUserName = "user1@mydomain.com"
}

resource "azuread_application" "test_registration" {
    name                       = "${var.azuread_appname}"
    homepage                   = "${var.azuread_homepage}"
    identifier_uris            = "${var.azuread_identifier_uris}"
    reply_urls                 = "${var.azuread_reply_urls}"
    available_to_other_tenants = "${var.azuread_available_to_other_tenants}"
    oauth2_allow_implicit_flow = "${var.azuread_oauth2_allow_implicit_flow}"
    approles                   = ["${data.azuread_application_group_role.myCustomGroupRole}, ${data.azuread_application_group_role.myCustomUserRole}]
}

```

### References

https://docs.microsoft.com/en-us/powershell/module/azuread/new-azureadgroupapproleassignment?view=azureadps-2.0
https://docs.microsoft.com/en-us/powershell/module/azuread/new-azureaduserapproleassignment?view=azureadps-2.0
https://docs.microsoft.com/en-us/powershell/module/azuread/set-azureadapplication?view=azureadps-2.0

* #0000

### PR DESCRIPTION
**Pull request by steve-hawkins**
this should resolve #75

### CODE REVIEW

File: azuread/resource_application.go [hunk 1]
```diff
@@ -300,6 +353,18 @@ func resourceApplicationUpdate(d *schema.ResourceData, meta interface{}) error {
 		properties.RequiredResourceAccess = expandADApplicationRequiredResourceAccess(d)
 	}
 
+	if d.HasChange("app_role") {
+		var disableAppRoles = true
```
Comment (katbyte) [inline azuread/resource_application.go]: I'm not sure the logic here, could you explain it in a comment?

File: azuread/resource_application.go [hunk 2]
```diff
@@ -5,7 +5,7 @@ import (
 	"log"
 
 	"github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac"
-
+	"github.com/hashicorp/go-uuid"
```
[the remark this answers is not in the document]
Comment (steve-hawkins) [reply azuread/resource_application.go]: done

File: azuread/resource_application.go [hunk 3]
```diff
@@ -198,13 +248,16 @@ func resourceApplicationCreate(d *schema.ResourceData, meta interface{}) error {
 		}
 	}
 
+	var disableAppRoles = false
```
[the remark this answers is not in the document]
Comment (steve-hawkins) [reply azuread/resource_application.go]: this has been removed and the logic improved also with comments

File: website/docs/r/application.html.markdown [hunk 4]
```diff
@@ -88,10 +102,24 @@ The following arguments are supported:
...
-* `id` - (Required) The unique identifier for one of the `OAuth2Permission` or `AppRole` instances that the resource application exposes. 
+* `id` - (Required) The unique identifier for one of the `OAuth2Permission` or `AppRole` instances that the resource application exposes.
 
 * `type` - (Required) Specifies whether the id property references an `OAuth2Permission` or an `AppRole`. Possible values are `Scope` or `Role`.
 
+---
+
+`app_role` supports the following:
+
+* `allowed_member_types` - (Required) Specifies whether this app role definition can be assigned to users and groups by setting to `User`, or to other applications (that are accessing this application in daemon service scenarios) by setting to `Application`, or to both.
```
[the remark this answers is not in the document]
Comment (steve-hawkins) [reply website/docs/r/application.html.markdown]: added thanks for spotting this

File: azuread/resource_application.go [hunk 5]
```diff
@@ -528,3 +596,70 @@ func flattenADApplicationOauth2Permissions(in []interface{}) []map[string]interf
...
+		}
+		var appRoleAllowedMemberTypes []string
+		for _, appRoleAllowedMemberType := range appRole["allowed_member_types"].(*schema.Set).List() {
+			appRoleAllowedMemberTypes = append(appRoleAllowedMemberTypes, appRoleAllowedMemberType.(string))
+		}
+		appRoleDescription := appRole["description"].(string)
+		appRoleDisplayName := appRole["display_name"].(string)
+		appRoleIsEnabled := appRole["is_enabled"].(bool)
+		if disable {
+			appRoleIsEnabled = false
```
[the remark this answers is not in the document]
Comment (steve-hawkins) [reply azuread/resource_application.go]: this has been removed and the logic improved also with comments

File: azuread/resource_application.go (hunk 1 above)
Comment (steve-hawkins) [reply azuread/resource_application.go]: this has been removed and the logic improved also with comments

File: azuread/resource_application.go [hunk 6]
```diff
@@ -528,3 +596,70 @@ func flattenADApplicationOauth2Permissions(in []interface{}) []map[string]interf
 
 	return result
 }
+
+func expandADApplicationAppRoles(in []interface{}, disable bool) *[]graphrbac.AppRole {
+	var results []graphrbac.AppRole
+
+	for _, appRoleRaw := range in {
+		appRole := appRoleRaw.(map[string]interface{})
+		appRoleID := appRole["id"].(string)
```
[the remark this answers is not in the document]
Comment (steve-hawkins) [reply azuread/resource_application.go]: id is indeed computed, but it is also required when updating or deleting an app role as they need to be disabled first

File: azuread/resource_application.go [hunk 7]
```diff
@@ -198,13 +248,16 @@ func resourceApplicationCreate(d *schema.ResourceData, meta interface{}) error {
...
+	var disableAppRoles = false
+
 	properties := graphrbac.ApplicationCreateParameters{
 		AdditionalProperties:    make(map[string]interface{}),
 		DisplayName:             &name,
 		IdentifierUris:          tf.ExpandStringSlicePtr(identUrls.([]interface{})),
 		ReplyUrls:               tf.ExpandStringSlicePtr(d.Get("reply_urls").(*schema.Set).List()),
 		AvailableToOtherTenants: p.Bool(d.Get("available_to_other_tenants").(bool)),
 		RequiredResourceAccess:  expandADApplicationRequiredResourceAccess(d),
+		AppRoles:                expandADApplicationAppRoles(d.Get("app_role").(*schema.Set).List(), disableAppRoles),
```
[the remark this answers is not in the document]
Comment (steve-hawkins) [reply azuread/resource_application.go]: sorry I'm pretty new to golang, do you have any examples in this or other providers I can follow?

File: azuread/resource_application.go [hunk 8]
```diff
@@ -178,6 +177,56 @@ func resourceApplication() *schema.Resource {
 				},
 			},
 
+			"app_role": {
```
Comment (katbyte) [inline azuread/resource_application.go]: Could we move this property to under type?

File: azuread/resource_application.go [hunk 9]
```diff
@@ -207,6 +256,10 @@ func resourceApplicationCreate(d *schema.ResourceData, meta interface{}) error {
 		RequiredResourceAccess:  expandADApplicationRequiredResourceAccess(d),
 	}
 
+	if v, ok := d.GetOk("app_role"); ok {
+		properties.AppRoles = expandADApplicationAppRoles(v.(*schema.Set).List())
```
Comment (katbyte) [inline azuread/resource_application.go]: Basically we would change this to:
```suggestion
	AppRoles: expandADApplicationAppRoles(d.Get("app_role"))
```

and move it up into the initializer above

File: azuread/resource_application.go [hunk 10]
```diff
@@ -297,6 +350,23 @@ func resourceApplicationUpdate(d *schema.ResourceData, meta interface{}) error {
 		properties.RequiredResourceAccess = expandADApplicationRequiredResourceAccess(d)
 	}
 
+	if d.HasChange("app_role") {
+		// if the app role already exists then it must be disabled
+		// with no other changes before it can be edited or deleted
+		var appRolesProperties graphrbac.ApplicationUpdateParameters
+		oldAppRolesRaw, newAppRolesRaw := d.GetChange("app_role")
+		appRolesProperties.AppRoles = expandADApplicationAppRoles(oldAppRolesRaw.(*schema.Set).List())
```
Comment (katbyte) [inline azuread/resource_application.go]: This doesn't account for if the app role has changed on the azure side. Would it make more sense to do a client.Get() to get the real state of the world here so the disablement won't fail?

File: azuread/resource_application.go [hunk 11]
```diff
@@ -178,6 +177,56 @@ func resourceApplication() *schema.Resource {
...
+
+						"allowed_member_types": {
+							Type:     schema.TypeSet,
+							Required: true,
+							MinItems: 1,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+								ValidateFunc: validation.StringInSlice(
+									[]string{"User", "Application"},
+									false, // force case sensitivity
```
Comment (katbyte) [inline azuread/resource_application.go]:
```suggestion
									false,

File: azuread/resource_application.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (katbyte) [reply azuread/resource_application.go]: ta

File: azuread/resource_application.go [hunk 12]
```diff
@@ -517,3 +590,75 @@ func flattenADApplicationOauth2Permissions(in *[]graphrbac.OAuth2Permission) []m
 
 	return result
 }
+
+func expandADApplicationAppRoles(input []interface{}) *[]graphrbac.AppRole {
```
Comment (katbyte) [inline azuread/resource_application.go]: Then this becomes:
```suggestion
func expandADApplicationAppRoles(i interface{}) *[]graphrbac.AppRole {
	input := i.(*schema.Set).List()
```

Comment (steve-hawkins) [reply azuread/resource_application.go]: thanks for your help with this, I have updated with your suggestions

File: azuread/resource_application.go (hunk 11 above)
Comment (steve-hawkins) [reply azuread/resource_application.go]: sorted

File: azuread/resource_application.go (hunk 9 above)
Comment (steve-hawkins) [reply azuread/resource_application.go]: sorted

File: azuread/resource_application.go (hunk 8 above)
Comment (steve-hawkins) [reply azuread/resource_application.go]: sorted

File: azuread/resource_application.go (hunk 10 above)
Comment (steve-hawkins) [reply azuread/resource_application.go]: I'll take a look into this

Comment (steve-hawkins) [reply azuread/resource_application.go]: think that should be sorted now

**Pull request merged**